### PR TITLE
Feature/102 integrate ollama llms

### PR DESCRIPTION
**Pull request by mthongvanh**
This PR implements the Ollama Dart client. The client allows consumers to interact with the Ollama API server by adding support for prompt response generation and several other Ollama model management API calls.

It provides the basis for resolving issue #102 Intergrate Ollama LLMs

The main dependencies are similar to the openai_dart package:
-   fetch_client: ^1.0.2
-   freezed_annotation: ^2.4.1
-   http: ^1.1.0
-   json_annotation: ^4.8.1
-   meta: ^1.7.0

 @davidmigloz

### CODE REVIEW

File: packages/ollama_dart/oas/ollama-curated.yaml [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+              schema:
+                $ref: '#/components/schemas/PushResponse'
+
+  /embeddings:
+    post:
+      tags:
+      - model management
+      summary: Generate embeddings from a model
+      description: Generate embeddings from a model
+      operationId: postEmbedding
```
Comment (davidmigloz) [inline packages/ollama_dart/oas/ollama-curated.yaml:202]: Maybe `generateEmbedding` is a better name

File: packages/ollama_dart/test/ollama_dart_test.dart [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+    tearDown(() {
+      client.endSession();
+    });
+
+    test('Ollama Client retrieves a list of local models', () async {
+      final TagResponse tags = await client.listTags();
+      expect(tags.models, isNotEmpty);
+    });
+
+    test('Ollama Client generates a response to a prompt', () async {
```
Comment (davidmigloz) [inline packages/ollama_dart/test/ollama_dart_test.dart:25]: We can add some more tests the completions API.

File: packages/ollama_dart/lib/src/client.dart [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    super.baseUrl,
+    final Map<String, String>? headers,
+    final http.Client? client,
+  }) : super(
+          headers: headers ?? const {},
+          client: client ?? createDefaultHttpClient(),
+        );
+
+  /// Host used when not provided in the class constructor
+  final defaultHost = 'http://localhost:11434/api';
```
Comment (davidmigloz) [inline packages/ollama_dart/lib/src/client.dart:32]: This one we don't really need it, if the user provides a different `baseUrl` the `makeRequestStream()` handles it internally.

File: packages/ollama_dart/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  build_runner: ^2.4.6
+  freezed: ^2.4.5
+  json_serializable: ^6.7.1
+#  openapi_spec: ^0.7.8
+  openapi_spec:
+    git:
+      url: https://github.com/davidmigloz/openapi_spec
+      ref: langchain
+  test: ^1.24.3
+  mockito: ^5.4.3
```
Comment (davidmigloz) [inline packages/ollama_dart/pubspec.yaml:35]: I don't think `mockito` is used

**Comment by davidmigloz** [comment]
Hey @mthongvanh!

I've reviewed the completions and embeddings part so far. Looking pretty good!

I've pushed a commit resolving all the conversations I opened.

I'll review the rest of the endpoints this evening. I think we can publish the first version today 🚀

Thanks again for your work 🙌

File: packages/ollama_dart/pubspec.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+name: ollama_dart
+description: Dart Client for the Ollama API
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/ollama_dart/pubspec.yaml]: updated

File: packages/ollama_dart/oas/ollama-curated.yaml [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+info:
+  title: Ollama API
+  description: API Spec for Ollama API server
+  version: 0.1.9
+
+servers:
+  - url: http://localhost:11434/api
+    description: local server
+  - url: http://localhost:11434/api
+    description: public server
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/ollama_dart/oas/ollama-curated.yaml]: updated

File: packages/ollama_dart/oas/ollama-curated.yaml [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+tags:
+  - name: inference
+    description: Prompt AI to generate output
+paths:
+  /generate:
+    post:
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/ollama_dart/oas/ollama-curated.yaml:20]: updated

File: packages/ollama_dart/oas/ollama-curated.yaml [hunk 8]
```diff
@@ -0,0 +1,624 @@
...
+          enum:
+            - creating system layer
+            - parsing modelfile
+            - success
+          example: parsing modelfile
+    
+    Options:
+      type: object
+      description: additional model parameters listed in the documentation for the Modelfile such as temperature
+      properties:
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/ollama_dart/oas/ollama-curated.yaml:265]: added

File: packages/ollama_dart/test/ollama_dart_test.dart [hunk 9]
```diff
@@ -0,0 +1,488 @@
+// ignore_for_file: avoid_print
+
+// import 'package:http/http.dart';
+import 'package:ollama_dart/ollama_dart.dart';
+import 'package:test/test.dart';
+
+void main() {
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/ollama_dart/test/ollama_dart_test.dart:5]: created separate files for completions and embeddings tests so far

File: packages/ollama_dart/oas/ollama-curated.yaml [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+        template:
+          type: string
+          description: the full prompt or prompt template (overrides what is defined in the Modelfile)
+        context:
+          type: array
+          description: the context parameter returned from a previous request to /generate, this can be used to keep a short conversational memory
+          items:
+            type: integer
+        stream:
+          type: boolean
```
[the remark this answers is not in the document]
Comment (davidmigloz) [reply packages/ollama_dart/oas/ollama-curated.yaml]: updated

File: packages/ollama_dart/oas/ollama-curated.yaml (hunk 1 above)
Comment (davidmigloz) [reply packages/ollama_dart/oas/ollama-curated.yaml]: updated

File: packages/ollama_dart/test/ollama_dart_test.dart (hunk 2 above)
Comment (davidmigloz) [reply packages/ollama_dart/test/ollama_dart_test.dart]: added some more tests cases

File: packages/ollama_dart/lib/src/client.dart (hunk 3 above)
Comment (davidmigloz) [reply packages/ollama_dart/lib/src/client.dart]: removed

File: packages/ollama_dart/pubspec.yaml (hunk 4 above)
Comment (davidmigloz) [reply packages/ollama_dart/pubspec.yaml]: removed

**Comment by mthongvanh** [comment]
great! thanks for cleaning up the PR.  i just found the generate query parameters in the documentation, so i can add those at a later date.

**Comment by davidmigloz** [comment]
I've finished reviewing the rest of the PR. It's good to go now 🚀

The only missing endpoint is the one to create a new blob. We need to do some fixes in the generator to support that case. But I guess it's not that important anyway.

**Review by davidmigloz — APPROVED** [review]

**Pull request merged**